### DISCUSSION TITLE
Internal error occurred during message handling / EIPE

### DISCUSSION THREAD

**Discussion by theogravity** [enisdenjo/graphql-ws#367] [Q&A] [answered]
`graphql-ws` sometimes prints out the following message:

https://github.com/enisdenjo/graphql-ws/blob/master/src/use/ws.ts#L140

```
Internal error occurred during message handling. Please check your implementation. Error: write EPIPE
```

Based on reading about `write EPIPE` errors, it seems that the client has disconnected during this phase:

https://github.com/enisdenjo/graphql-ws/blob/master/src/server.ts#L590

In our logging infra, it goes across many lines and looks like this:

<img width="788" alt="CleanShot 2022-06-10 at 16 12 15@2x" src="https://user-images.githubusercontent.com/855434/173161404-41abd6e6-4b78-488b-bb56-1d7178e76df7.png">

- Should we be able to define the logger we want to use for errors so we can parse them to how we want?
- What's the best way to catch this kind of error when it happens and where should it be done?
- Is it something on our side or really due to the client disconnecting during processing?

I wonder if my issue is the same as https://github.com/enisdenjo/graphql-ws/discussions/333

We do have a lot of complex logic that could potentially throw in our `subscription` handlers...

**Comment by enisdenjo** [comment, marked as answer]
> Should we be able to define the logger we want to use for errors so we can parse them to how we want?

I'd recommend implementing your own server instead. You can then handle the errors however you see fit. For a quick start, please consult the ["Server usage with ws" recipe](https://github.com/enisdenjo/graphql-ws#ws).

> What's the best way to catch this kind of error when it happens and where should it be done?

All possible errors that could be thrown during operation execution can be handled using the [onMessage callback argument](https://github.com/enisdenjo/graphql-ws/blob/77682cd3d9bac748b5b24e2f8cf88e1d98c7a448/README.md?plain=1#L962-L972). Read more about it in [the documentation](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.WebSocket.md#onmessage).

> Is it something on our side or really due to the client disconnecting during processing?

I've never faced this error myself, so I cannot shine more light on this question. If it is indeed #333, my recommendation is to just handle the errors properly in the `subscribe` handlers.
